src/metabase/notification/send.clj

```diff
@@ -188,12 +181,89 @@
       (prometheus/observe! :metabase-notification/total-duration-ms {:payload-type payload_type} total-time))
     nil))
 
+(defprotocol NotificationQueueProtocol
+  "Protocol for notification queue implementations."
+  (put-notification!  [this notification] "Add a notification to the queue. If a notification with the same id is already in the queue, replace it.")
+  (take-notification! [this]              "Take the next notification from the queue, blocking if none available."))
+
+(deftype ^:private NotificationQueue
+         [^java.util.LinkedList ids-list
+          ^java.util.concurrent.ConcurrentHashMap id->notification
+          ^java.util.concurrent.locks.ReentrantLock queue-lock
```

Comment:
Why do we need a re-entrant lock? I don't see where we use recursion.

Reply:
om, I'm not a lock expert but I think that's the popular one that supports condition?

Reply:
Ah, and intrinsic locks don't. Gotcha.